controllers/machinehealthcheck_controller_test.go

```diff
@@ -949,7 +945,167 @@ func TestMachineHealthCheck_Reconcile(t *testing.T) {
 			return
 		}).Should(Equal(1))
 
-		// Calculate how many Machines have been remediated.
+		// Transition the node back to healthy.
+		node = nodes[0]
+		nodePatch = client.MergeFrom(node.DeepCopy())
+		node.Status.Conditions = []corev1.NodeCondition{
+			{
+				Type:               corev1.NodeReady,
+				Status:             corev1.ConditionTrue,
+				LastTransitionTime: metav1.NewTime(time.Now().Add(-10 * time.Minute)),
+			},
+		}
+		g.Expect(testEnv.Status().Patch(ctx, node, nodePatch)).To(Succeed())
+
+		// Make sure the status matches.
+		g.Eventually(func() *clusterv1.MachineHealthCheckStatus {
+			err := testEnv.Get(ctx, util.ObjectKey(mhc), mhc)
+			if err != nil {
+				return nil
+			}
+			return &mhc.Status
+		}).Should(MatchMachineHealthCheckStatus(&clusterv1.MachineHealthCheckStatus{
+			ExpectedMachines:   1,
+			CurrentHealthy:     1,
+			ObservedGeneration: 1,
+			Targets:            targetMachines},
+		))
+
+		// Calculate how many Machines have health check succeeded = false.
+		g.Eventually(func() (unhealthy int) {
+			machines := &clusterv1.MachineList{}
+			err := testEnv.List(ctx, machines, client.MatchingLabels{
+				"selector": mhc.Spec.Selector.MatchLabels["selector"],
+			})
+			if err != nil {
+				return -1
+			}
+
+			for i := range machines.Items {
+				if conditions.IsFalse(&machines.Items[i], clusterv1.MachineHealthCheckSuccededCondition) {
+					unhealthy++
+				}
+			}
+			return
+		}).Should(Equal(0))
+
+		// Make sure the Remeditaion Request is deleted.
+		g.Eventually(func() bool {
+			remediationTmpl := remediationTemplate.DeepCopy()
+			err := testEnv.Get(ctx, util.ObjectKey(infraRemediationTmpl), remediationTmpl)
+			return apierrors.IsNotFound(err)
+		}).Should(BeTrue())
```

Comment:
When I try to verify the deletion of the Remediation Request after Node is back to healthy, the test if failing. I assumed that if Get returns IsNotFound error our boolean should be true. 

But I am not totally sure if the right object is tested here. In the beginning of the test the Remediation Template is created and based on that Template controller should create the Remediation Request. Maybe the existence of template is tested instead of testing existence of Remediation Request?

Reply:
Yes this is testing that the template exists, not the actual object. We need to create a new unstructured object with the expected name / GVK and try to get that one instead

Reply:
When I print out the error from the `testEnv.Get`, I see the following: `no kind is registered for the type v1.ObjectReference in scheme "pkg/runtime/scheme.go:101`. I believe Vince is correct, we need to get the actual remediation request object and assert on that.